src/Groups/matrices/FiniteFormOrthogonalGroup.jl

```diff
@@ -0,0 +1,843 @@
+  ind, val = _block_indices_vals(G, p)
+  # add a virtual even block at the end
+  push!(ind, n+1)
+  push!(val, val[end] - 1)
+
+  # block diagonal contribution
+  gens = typeof(G)[]
+  for k in 1:length(ind)-1
+    i1 = ind[k]
+    i2 = ind[k+1]
+    Gk = divexact(G[i1:i2-1,i1:i2-1], p^val[k])
+    Gk_inv = inv(Gk)
+    ni = i2 - i1
+    Ek = identity_matrix(R, ni)
+    Zk = zero_matrix(R, ni, ni)
+    if p == 2 && a == 1
+      genskmodp = _solve_X_ker(Zk, zeros(R, ni), diagonal(Gk_inv))
+      gensk = [change_base_ring(R,lift(g)) for g in genskmodp]
+    else
+      # basis for the space of anti-symmetric matrices
+      gensk = typeof(G)[]
+      for i in 1:ni
+        for j in 1:i-1
+          gk = deepcopy(Zk)
+          gk[i,j] = 1
+          gk[j,i] = -1
+          push!(gensk, gk)
+        end
+      end
+    end
+    for h in gensk
+      g = deepcopy(E)
+      g[i1:i2-1, i1:i2-1] = Ek + p^a*h*Gk_inv
+      push!(gens, g)
+    end
+  end
+  # generators below the block diagonal.
+  for i in 1:n
+    for j in 1:i
+      g = deepcopy(E)
+      g[i,j] = p^a
+      flag = true
+      for k in 1:(length(ind)-1)
+        # exclude the block diagonal entries
+        if ind[k] <= i < ind[k+1] && ind[k] <= j < ind[k+1]
+          flag = false
+          break
+        end
+      end
+      if flag
+        push!(gens, g)
+      end
+    end
+  end
+  return gens
+end
+
+#     r"""
+#     Return the transformation to normal form.
+#
+#     INPUT:
+#
+#     - ``G`` -- `p`-adic symmetric matrix.
+#
+#     OUTPUT:
+#
+#     - ``D`` -- the normal form
+#     - ``B`` -- a transformation matrix
+#
+function _normal(G::Union{fmpz_mod_mat, nmod_mat}, p)
+  if p == 2
+    D1, B1 = _jordan_2_adic(G)
+    D2, B2 = _normalize(D1, p)
+    D3, B3 = _two_adic_normal_forms(D2, p)
+    B = B3 * B2 * B1
+    return D3, B
+  else
+    D1, B1 = _jordan_odd_adic(G, p)
+    D2, B2 = _normalize(D1, p)
+    B = B2 * B1
+    return D2, B
+  end
+end
+
+function lift(x::AbstractAlgebra.Generic.MatSpaceElem{gfp_fmpz_elem})
+  return map_entries(lift , x)
+end
+
+#     Return generators of the orthogonal group modulo `p` for odd `p`.
+#
+#     `G` is the gram matrix of the bilinear form.
+function _gens_form(G::Union{fmpz_mod_mat, nmod_mat}, form_constructor, p)
+    R = base_ring(G)
+    if ncols(G) == 0
+        return typeof(G)[]
+    end
+    F = GF(p, 1)
+    Gf = change_base_ring(F, lift(G))
+    FF = GF(p)
+    q = form_constructor(Gf)
+    Oq = isometry_group(q)
+    gensOq = gens(Oq)
+    gensOq = [change_base_ring(FF,matrix(g)) for g in gensOq]
+    gensOq = [change_base_ring(R,lift(g)) for g in gensOq]
+
+    return gensOq
+end
+
+function _gens_qf(G::Union{fmpz_mod_mat, nmod_mat}, p)
+  Gq = deepcopy(G)
+  if p == 2
+    r = ncols(G)
+    for i in 1:r
+      for j in 1:i-1
+        Gq[i,j] = 0
+      end
+      Gq[i,i] = divexact(Gq[i,i],2)
+    end
+  end
+  gensOq =  _gens_form(Gq::Union{fmpz_mod_mat, nmod_mat}, quadratic_form, p)
```

Comment:
```suggestion
  gensOq =  _gens_form(Gq, quadratic_form, p)
```
Should not be necessary.

Reply:
I think it might be useful to keep for documentation purposes.